bootenv.go

```diff
@@ -33,22 +34,39 @@ func NewEnvironment(cmd Commander) *uBootEnv {
 
 func (e *uBootEnv) ReadEnv(names ...string) (BootVars, error) {
 	getEnvCmd := e.Command("fw_printenv", names...)
-	return getOrSetEnvironmentVariable(getEnvCmd)
+	return getEnvironmentVariable(getEnvCmd)
 }
 
 func (e *uBootEnv) WriteEnv(vars BootVars) error {
-	//TODO: try to make this atomic later
+	// Make environment update atomic by using fw_setenv "-s" option.
+	setEnvCmd := e.Command("fw_setenv", "-s", "-")
+	pipe, err := setEnvCmd.StdinPipe()
+	if err != nil {
+		log.Errorln("Could not set up pipe to fw_setenv command: ", err)
+		return err
+	}
+	err = setEnvCmd.Start()
+	if err != nil {
+		log.Errorln("Could not execute fw_setenv: ", err)
+		return err
+	}
 	for k, v := range vars {
-		setEnvCmd := e.Command("fw_setenv", k, v)
-		if _, err := getOrSetEnvironmentVariable(setEnvCmd); err != nil {
-			log.Error("Error setting U-Boot variable: ", err)
+		_, err = fmt.Fprintf(pipe, "%s %s\n", k, v)
+		if err != nil {
+			log.Error("Error while setting U-Boot variable: ", err)
 			return err
 		}
 	}
+	pipe.Close()
```

Comment:
You could also close executing a command inside a function and call defer there without worrying about calling `Close()` in each return path. But this way it will work as well.